ambry-router/src/main/java/com.github.ambry.router/CryptoJobHandler.java

```diff
@@ -30,28 +34,54 @@
  * and any new jobs submitted after close will be ignored.
  */
 class CryptoJobHandler implements Closeable {
+  private final long cryptoJobTimeoutMs;
   private final AtomicBoolean enabled = new AtomicBoolean(false);
   private static final GeneralSecurityException CLOSED_EXCEPTION =
       new GeneralSecurityException("CryptoJobHandler closed");
-  private ExecutorService scheduler;
+  private static final GeneralSecurityException CRYPTO_JOB_TIMEOUT_EXCEPTION =
+      new GeneralSecurityException("CryptoJob timeout", new TimeoutException());
+  private ExecutorService executor;
+  private ConcurrentHashMap<CryptoJob, Long> pendingCryptoJobs = new ConcurrentHashMap<>();
```

Comment:
`final`?

Reply:
Also why is this a `ConcurrentHashMap`? Are additions/removals done across threads (it doesn't look like that in the code)? If that is not the case, it is better to make this a `TreeMap` sorted on the `long` since you are looking for expiry. 

Giving it some thought, isn't a queue the better datastructure to use here since you want it sorted by time? Using a blocking queue and a `poll()` should give you all the semantics you need.